manual: architecture overview (#7066)
NixOS/nix: doc/manual/src/architecture/architecture.md

|                 [ configuration language  ]      |            |
|                              |                   |            |
| +----------------------------|-------------------V----------+ |
| | store                 evaluates to                        | |
| |                            |                              | |
| |            referenced by   V       builds                 | |
| | [ build input ] ---> [ build plan ] ---> [ build result ] | |
| |                                                           | |
| +-----------------------------------------------------------+ |
+---------------------------------------------------------------+
```

At the top is the [command line interface](../command-ref/command-ref.md), translating from invocations of Nix executables to interactions with the underlying layers.

Below that is the [Nix language](../language/index.md), a [purely functional] configuration language.
It is used to compose expressions which ultimately evaluate to self-contained *build plans*, used to derive *build results* from referenced *build inputs*.

[purely functional]: https://en.m.wikipedia.org/wiki/Purely_functional_programming

> Command line interface and Nix language are what users interact with most.

Author: ```suggestion
The command line interface and the Nix language are what users interact with most.
```

> **Note**
> The Nix language itself does not have a notion of *packages* or *configurations*.
> As far as we are concerned here, the inputs and results of a build plan are just data.

> Underlying these is the [Nix store](../glossary.md#gloss-store), a mechanism to keep track of build plans, data, and references between them.
> It can also execute build plans to produce new data.

Reviewer: ```suggestion
Underlying the command line interface and the Nix language is the [Nix store](../glossary.md#gloss-store), a mechanism to keep track of build plans, data, and references between them.
It can also execute build plans to produce new data.
```

My tiny brain already forgot what came before... 😅

Author: Probably has nothing to do with your brain specifically.

# Architecture

This chapter describes how Nix works.
It should help users understand why Nix behaves as it does, and it should help developers understand how to modify Nix and how to write similar tools.

## Overview

Nix consists of [hierarchical layers].

[hierarchical layers]: https://en.m.wikipedia.org/wiki/Multitier_architecture#Layers

```
+---------------------------------------------------------------+
| Nix             .-------------------------.                   |
|                 | commmand line interface |------.            |
|                 '-------------------------'      |            |
|                              |                   |            |

> |                          evaluates               |            |

Reviewer 2: This suggests that the CLI layer evaluates the Nix language, but that's the job of the `libexpr` layer.

|                              |                manages         |
|                              V                   |            |
|                 .-------------------------.      |            |
|                 | configuration language  |      |            |
|                 '-------------------------'      |            |
|                              |                   |            |
|                         evaluates to             |            |
|                              |                   V            |
| +----------------------------|------------------------------+ |
| | store                      |                              | |
| |            referenced by   V       builds                 | |
| | .-------------.      .------------.      .--------------. | |
| | | build input |----->| build plan | ---->| build result | | |
| | '-------------'      '------------'      '--------------' | |
| +-----------------------------------------------------------+ |

> +---------------------------------------------------------------+

Reviewer 3: Not sure if I'm misunderstanding, but `referenced by` seems to go in the opposite direction of the arrow.

Author: Build inputs are referenced by build tasks.

Yes, the arrow directions can be ambiguous. As we discussed in yesterday's maintainer meeting, the point of these diagrams is to convey the right general idea in order to help shape the first approximation of a mental
model, not to be mathematically precise.

There is not enough space in such a diagram and not enough capacity on the reader's end to capture all the details, and this introductory document is not meant to hold that information.

The necessary details will follow follow-up PRs adding sub-chapters, which will be linked from the respective terms.

Reviewer 3: Can’t we just invert the arrow and say “build inputs reference build tasks”?

Reviewer 3: Alright, just wanted to double check. Consider my question answered. Thanks!

> +---------------------------------------------------------------+

Reviewer 2: It would be good if this picture includes the names of the components: `libstore`, `libexpr`, `libcmd / src/nix`, ...

```

At the top is the [command line interface](../command-ref/command-ref.md), translating from invocations of Nix executables to interactions with the underlying layers.

Below that is the [Nix language](../language/index.md), the configuration language for Nix.
Its expressions ultimately evaluate to self-contained *build plans*, used to derive *build results* from referenced *build inputs*.

The command line interface and the Nix language are what users interact with most.

> **Note**
> The Nix language itself does not have a notion of *packages* or *configurations*.
> As far as we are concerned here, the inputs and results of a build plan are just data.

Underlying the command line interface and the Nix language is the [Nix store](../glossary.md#gloss-store), a mechanism to keep track of build plans, data, and references between them.

> It can also execute build plans to produce new data.

Reviewer 2: "It" here seems to refer to the Nix store, which cannot be said to execute build plans.

> A build plan is a series of *build tasks*.

Reviewer 3: I find it a bit counterintuitive that the first term introduced is "build task", only then for the term to be redefined in a "boxed paragraph".

Also, by introducing this term ("build task"), I think you're increasing the cognitive load, as it is not going to be useful in the broader Nix context.

My suggesiton is instead 

```suggestion
A build plan is a series of "build tasks", which in Nix are called [derivations](../glossary#gloss-derivation).
```

Reviewer 2: Yes, maybe it's better to use "derivation" instead of "build task", which is not a term that appears in the Nix source.

Author: We still don't have a definition of what exactly a derivation even is. I'd argue the term derivation should be reserved for Nix's specific implementation of build tasks, whatever it is we define it to be - store derivation or the Nix language concept.
